### ISSUE TITLE
SC2312 should be a little smarter

### ISSUE THREAD

**Issue by oliv3r** [koalaman/shellcheck#2775]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2312
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

With SC2312, I wonder
```sh
printf "$(cat '/etc/hostname')"
```
will yield the dreaded SC2312 error. Probably rightfully so. But in this particular case (as with most of these) I think that shellcheck should accept not just `|| true` to ignore the warning (or accept the fact), but as the rationale/exception says 'if you have other means.

While this will be a very slippery slope, I think
```sh
printf "$(cat '/etc/hostname' || echo 'Unknown')"
# and
printf "$(cat '/etc/hostname' || printf 'Unknown')"
```
I think are both valid choices in addition to `|| true`. I don't think shellcheck fails on `printf "$(echo 'test')"` does it? As echo and friends are considered to be okay to ignore return values (I'm not even sure lol).

I actually even wonder in pipelines, should not one final 'echo' cover it all?
```sh
printf "$(cat '/etc/hostname' | sed 's/match/remove/g' || echo 'Unknown')"
```
in the end, we care about the end-result of the subshell, not all intermediate steps ... (I know this can get very complex very fast).
